### ISSUE TITLE
Accessing Blob Storage contents from custom IoT Edge module

### ISSUE THREAD

**Issue by gri6507** [Azure/iotedge#2612]
I am creating an IoT Edge device with a custom IoT Edge Module which needs to get blob contents from an Azure Blob Storage account. Based on https://docs.microsoft.com/en-us/azure/iot-edge/how-to-deploy-blob#deploy-from-the-azure-portal, I updated my device deployment policy `content.modulesContent.$edgeAgent.properties.desired.modules` to include 
```json
"azureblobstorageoniotedge": {
    "settings": {
        "image": "mcr.microsoft.com/azure-blob-storage:latest",
        "createOptions": "{\"Env\":[\"LOCAL_STORAGE_ACCOUNT_NAME=blobstore\",\"LOCAL_STORAGE_ACCOUNT_KEY=<a key I generated>\"],\"HostConfig\":{\"Binds\":[\"/blobroot:/blobroot\"],\"PortBindings\":{\"11002/tcp\":[{\"HostPort\":\"11002\"}]}}}"
    },
    "type": "docker",
    "status": "running",
    "restartPolicy": "always",
    "version": "1.0"
}
```
and `content.modulesContent` to include
```json
"azureblobstorageoniotedge": {
    "properties.desired": {
        "deviceAutoDeleteProperties": {
            "deleteOn": false
        },
        "deviceToCloudUploadProperties": {
            "uploadOn": false,
            "cloudStorageConnectionString": "DefaultEndpointsProtocol=https;AccountName=<CloudBlobStorageAccountName>;AccountKey=<CloudBlobStorageAccountKey>;EndpointSuffix=core.windows.net"
        }
    }
}
```
Per instructions, on my edge hardware machine, I created the blobroot that the `azureblobstorageoniotedge` module will use

```bash
sudo mkdir /blobroot
sudo chown -R 11000:11000 /blobroot
sudo chmod -R 700 /blobroot
```
I redeployed my edge device with the new deployment policy. This added the `azureblobstorageoniotedge` module to my device and I can see that it's using the blobroot for something (the content of those files doesn't make sense to me)

```bash
sudo find /blobroot
/blobroot
/blobroot/Database
/blobroot/Database/LOCK
/blobroot/Database/LOG
/blobroot/Database/OPTIONS-000020
/blobroot/Database/OPTIONS-000018
/blobroot/Database/IDENTITY
/blobroot/Database/CURRENT
/blobroot/Database/000003.log
/blobroot/Database/MANIFEST-000004
/blobroot/BlockBlob
```

In my custom IoT Edge module I am using the `azure-storage-cpp` SDK to get blob contents. I developed and tested the code outside of the IoT Edge module environment and, when connecting with the connection string to the actual cloud blob storage, I can see the blob storage container and then download the desired blob from that container.

Of course, in order to run the same code as part of my custom IoT Edge module, per instructions, I changed the connection string to reference my local instance of blob storage at `http://azureblobstorageoniotedge:11002/blobstore` (see full connection string below). When my custom IoT Edge module runs, as evidenced by `iotedge logs azureblobstorageoniotedge -f`, it connects to `azureblobstorageoniotedge`, but there is no container! Listing all available containers within the `blob_client` shows nothing and simply downloading a blob fails.

What am I missing?

```c++
#include <was/storage_account.h>
#include <was/blob.h>

#include <iostream>


int download()
{
#if 1

  const utility::string_t storage_connection_string(U("DefaultEndpointsProtocol=https;AccountName=<CloudBlobStorageAccountName>;AccountKey=<CloudBlobStorageAccountKey>;EndpointSuffix=core.windows.net"));

#else

  const utility::string_t storage_connection_string(U("DefaultEndpointsProtocol=http;BlobEndpoint=http://azureblobstorageoniotedge:11002/blobstore;AccountName=blobstore;AccountKey=<a key I generated>"));

#endif

  const utility::string_t container_name(U("my-blob-container"));
  const std::string blob_path("path/to/my/blob");

  try 
  {
    azure::storage::cloud_storage_account storage_account = azure::storage::cloud_storage_account::parse(storage_connection_string);
    azure::storage::cloud_blob_client blob_client = storage_account.create_cloud_blob_client();

    std::cout << "listing client containers" << std::endl;
    for (auto&& item: blob_client.list_containers())
    {   
      std::cout << "found container " << item.name() << std::endl;
    }   

    std::cout << "Downloading blob: " << blob_path << std::endl;
    azure::storage::cloud_blob_container container = blob_client.get_container_reference(container_name);
    auto a_blob = container.get_blob_reference(U(blob_path));
    if (!a_blob.exists())
    {   
      throw std::runtime_error("Blob does not exist");
    }   
    a_blob.download_attributes();
    auto blob_properties = a_blob.properties();
    std::cout << U("Blob: ") << a_blob.uri().primary_uri().to_string() << " MD5: " << blob_properties.content_md5() << " ETag: " << blob_properties.etag() << std::endl;

    concurrency::streams::container_buffer<std::vector<uint8_t>> buffer;
    concurrency::streams::ostream output_stream(buffer);
    a_blob.download_to_stream(output_stream);
    std::vector<unsigned char>& data = buffer.collection();
    std::string blob_contents((char *)&data[0], buffer.size());

    std::cout << blob_contents << std::endl;

  }
  catch (const std::exception& e)
  {
      std::wcout << U("BLOB ERROR: ") << e.what() << std::endl;
      return 1;
  }

  return 0;
}
```

**Comment by mbialecka** [comment]
@gri6507 Have you created a container and a blob in the blob storage module? You can check logs or /blobroot/BlockBlob (a subdirectory is created for every container).

**Comment by gri6507** [comment]
@mbialecka - thank you for the prompt reply. Yes, the cloud blob storage account has the target container and the target blob path. As I pointed out in the original post, when I run the above C++ code on my local PC (not in IoT Edge Module) with the connection string directly to the Cloud blob storage account, then I can download the blob without problems.

The issue is that I cannot use the cloud blob storage connection string from with my custom IoT Edge module (I get SSL handshake error) and when I use the local blob storage account, I cannot download the blob in question.

My understanding of how the `azureblobstorageoniotedge` works is that it is essentially a mirror of the cloud side storage, but is accessible from within my custom IoT Edge module. So, I would expect that any blob that is in the cloud should also be accessible from the local `azureblobstorageoniotedge`

**Comment by mbialecka** [comment]
@gri6507 No, the module does not mirror cloud account. The module allows you to store blobs locally (on the device running the module) using the same APIs as cloud storage and it has an option to upload blobs from module to the cloud. The module never downloads blobs from the cloud.

**Comment by gri6507** [comment]
@mbialecka - Thank you for that clarification. In that case, can you please suggest the proper way to download cloud resident blobs into my IoT Edge module? 

Like I said, the `azure-storage-cpp` based code above can download a blob when running on my local PC (not in IoT Edge Module) with the connection string directly to the Cloud blob storage account. However, if I use the same code with the same connection string from within my IoT Edge module, I get an SSL Handshake Error. Perhaps I need to update my deployment policy to start my IoT Edge module with some other docker container settings? Here's the deployment policy snippet related to my module

```json
"MyIoTEdgeModule": {
    "version": "242",
    "type": "docker",
    "status": "running",
    "restartPolicy": "always",
    "settings": {
        "image": "<ACR URI to my module>",
        "createOptions": "{\"HostConfig\":{\"CapAdd\":[\"SYS_PTRACE\"], \"ExtraHosts\":[\"dockerhost:172.42.0.1\"]}}"
    }
}
```

**Comment by sudhirap** [comment]
@gri6507 Your scenario is about mirroring the blob namespace in Azure to the IoT Edge device and then accessing it through a local blob endpoint, correct? This is unsupported through the Blob Storage on IoT Edge module. The main support scenario is local storage of blobs and their further upload to Azure.

**Comment by gri6507** [comment]
@sudhirap - thank you for clarifying that my use case is not supported by the `azureblobstorageoniotedge` module. However, this is a real use case for my application. How does Microsoft recommend I download blobs from the cloud into my edge module?

**Comment by sudhirap** [comment]
If you are running on Linux please look into this link - https://docs.microsoft.com/en-us/azure/storage/blobs/storage-how-to-mount-container-linux

This provides a local file system namespace backed by Azure block blobs. However if you want a local blob namespace then please search online, I can’t think of any right away for Azure blobs.
Having said that is your scenario a sync one or a download one i.e. something changes in Azure just download the changed blob as per some policy and go back periodically rather than smartly tracking changes and caching them locally?

**Comment by gri6507** [comment]
@sudhirap, @mbialecka  - While my solution does run on Linux, using the Fuse File System approach you referenced in https://docs.microsoft.com/en-us/azure/storage/blobs/storage-how-to-mount-container-linux does not make sense because it forces each instance of the Edge device to have a **complete** copy of the Cloud resident blob storage. This is a **lot** of data, especially since only one file is needed. Also, that approach introduces additional problems with synchronicity between the instance of the one file of interest on the cloud and the local fuse filesystem.

My complete use case is as follows. The custom IoT Edge Module code is generic for a very large number of devices. However, each instance of the edge device with the custom module needs specific unique configuration information. This information is too large to fit into the module twin. Therefore, we are using the twin to reference a container+blob path in Cloud blob storage to allow each module to configure itself uniquely.

That being said, our cloud storage has a relatively large number of files in it, but **only one** file is needed by the edge module. There has to be a way to use the `azure-storage-cpp` SDK to do that. Like I said in the beginning, I wrote the code and tested it on my local PC (not in an edge module), so I know that works. However, when I run exactly the same code with exactly the same connection string to the cloud blob storage account, I get `SSL Handshake` errors.

In theory, the `mcr.microsoft.com/azure-blob-storage:latest` module from Microsoft has had to address this problem somehow. I know it is using a .NET version of the SDK, but it had to have done something to work around this `SSL Handshake` error. Where can I ask this question? Perhaps there is a GitHub repo for the `azure-blob-storage` module where I can see the code and ask my question?

**Comment by sudhirap** [comment]
Blob Storage module for IoT Edge supports upload scenarios not download. If all you are looking for is to download a single configuration file as a blob from cloud storage to the edge then yes Fuse is too heavy.
The Blob Storage module itself uses Blob Client SDK to upload to the cloud. You should be able to use the SDK to download blobs. Can you please take a look at https://docs.microsoft.com/en-us/azure/iot-edge/how-to-deploy-blob, specifically at cloudStorageConnectionString for the connection string to pass in to your blob client call? You can then use https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobserviceclient?view=azure-dotnet.

**Comment by gri6507** [comment]
@sudhirap - I am not using .NET, but rather the c++ Blob client called [azure-storage-cpp](https://azure.github.io/azure-storage-cpp/) . As you can see in my first post of this thread, following the [example code](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-c-plus-plus-how-to-use-blobs), I was able to write a simple test application to download a single blob file from my cloud blob storage account **on my local PC (not in an edge module)**

However, like I said above, when I use the exact same code in my IoT Edge module, the blob storage client reports `Error in SSL handshake`. Looking into it a bit deeper, the `azure-storage-cpp` C++ blob client SDK uses the `cpprestsdk` as the underlying REST API. This error comes from https://github.com/microsoft/cpprestsdk/blob/master/Release/src/http/client/http_client_asio.cpp#L1099

My question is why does my code, which works just file outside of the edge module, cause this failure when running inside the edge module? How do I fix it to work in the edge module?

**Comment by gri6507** [comment]
@sudhirap - I found the answer to my own question. I knew that my `azure-storage-cpp` based code works outside of the edge module runtime. After thinking about what may cause SSL Handshake problems inside the edge module, it occurred to me that my edge module, which is based on Ubuntu 18.04, does not have the `ca-certificates` package installed. Without that package, the runtime has no way to validate the root certificate for the blob storage account connection, and that's why I was getting the SSL Handshake error. 

All I had to do was update my Dockerfile for my module to `apt install ca-certificates` and everything now works!
